src/lib/utils/tree_hash.h

```diff
@@ -0,0 +1,269 @@
+/**
+ * Treehash logic used for hash-based signatures
+ * (C) 2023 Jack Lloyd
+ *     2023 Fabian Albert, René Meusel, Amos Treiber, Philippe Lieser - Rohde & Schwarz Cybersecurity GmbH
+ *
+ * Parts of this file have been adapted from https://github.com/sphincs/sphincsplus
+ *
+ * Botan is released under the Simplified BSD License (see license.txt)
+ */
+#ifndef BOTAN_TREE_HASH_H_
+#define BOTAN_TREE_HASH_H_
+
+#include <botan/exceptn.h>
+#include <botan/strong_type.h>
+#include <botan/internal/stl_util.h>
+
+#include <cstdint>
+#include <functional>
+#include <optional>
+#include <vector>
+
+namespace Botan {
+
+namespace concepts {
+
+template <typename T>
+concept tree_node = contiguous_container<T>;
+
+/**
+ * @brief An index of a node in a layer.
+ *
+ * This is a separate index for each layer.
+ * The left most node of a layer has the index 0.
+ */
+template <typename T>
+concept tree_node_index = strong_type_with_capability<T, EnableArithmeticWithPlainNumber>;
```

Comment:
Just an alternative suggestion, because `strong_type_with_capability` seems to be so verbose that we'll probably always wrap it into more specific concepts: Maybe not have `strong_type_with_capability` at all but formulate those as a conjunction:

```suggestion
template <typename T>
concept tree_node_index = is_strong_type<T>() && T::template has_capability<EnableArithmeticWithPlainNumber>;
```

Reply:
Do you think so? I would guess that it's not a rare use case. And since strong types are the only classes with Capabilities it is reasonable IMO to have a concept for this. What do you think?

Reply:
Don't have a super strong opinion on that. You have a point that it is more readable for sure.